test/torch/test_hook.py

```diff
@@ -11,194 +11,227 @@
+    alice = syft.VirtualWorker(id=f"alice{instance_id}", hook=hook, is_client_worker=False)
+
+    x = torch.Tensor([1, 2, 3])
+    native_method = getattr(x, f"native_{attr}")
+    expected = native_method(x)
+    x_ptr = x.send(bob).send(alice)
+    method = getattr(x_ptr, attr)
+    res_ptr = method(x_ptr)
+    res = res_ptr.get().get()
+    assert (res == expected).all()
+
+
+@pytest.mark.parametrize("attr", ["relu", "celu", "elu"])
+def test_hook_module_functional(attr):
+    hook = syft.TorchHook(torch, verbose=True)
+
+    instance_id = str(int(10e10 * random.random()))
+
+    bob = syft.VirtualWorker(id=f"bob{instance_id}", hook=hook, is_client_worker=False)
+
+    attr = getattr(F, attr)
+    x = torch.Tensor([1, -1, 3, 4])
+    expected = attr(x)
+    x_ptr = x.send(bob)
+    res_ptr = attr(x_ptr)
+    res = res_ptr.get()
+    assert (res == expected).all()
+
+
+@pytest.mark.parametrize("attr", ["relu", "celu", "elu"])
+def test_functional_same_in_both_imports(attr):
+    """This function tests that the hook modifies the behavior of
+    torch.nn.function regardless of the import namespace
+    """
+    fattr = getattr(F, attr)
+    tattr = getattr(torch.nn.functional, attr)
+    x = torch.Tensor([1, -1, 3, 4])
+    assert (fattr(x) == tattr(x)).all()
+
+
+def test_hook_tensor():
+    hook = syft.TorchHook(torch, verbose=True)
+
+    instance_id = str(int(10e10 * random.random()))
+
+    bob = syft.VirtualWorker(id=f"bob{instance_id}", hook=hook, is_client_worker=False)
+
+    x = torch.tensor([1.0, -1.0, 3.0, 4.0], requires_grad=True)
+    x.send(bob)
+    x = torch.tensor([1.0, -1.0, 3.0, 4.0], requires_grad=True)[0:2]
+    x.send(bob)
+
+
+def test_properties():
+    x = torch.Tensor([1, -1, 3, 4])
+    assert x.is_wrapper is False
+
+
+def test_signature_cache_change():
+    """Tests that calls to the same method using a different
+    signature works correctly. We cache signatures in the
+    hook.build_hook_args_function dictionary but sometimes they
+    are incorrect if we use the same method with different
+    parameter types. So, we need to test to make sure that
+    this cache missing fails gracefully. This test tests
+    that for the .div(tensor) .div(int) method."""
+
+    x = torch.Tensor([1, 2, 3])
+    y = torch.Tensor([1, 2, 3])
+
+    z = x.div(y)
+    z = x.div(2)
+    z = x.div(y)
+
+    assert True
+
+
+def test_parameter_hooking():
+    """Test custom nn.Module and parameter auto listing in m.parameters()"""
+
+    class MyLayer(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.some_params = torch.nn.Parameter(torch.tensor([5.0]))
+
+    m = MyLayer()
+    out = list(m.parameters())
+    assert len(out) == 1
+    assert out[0] == m.some_params
+
+
+def test_torch_module_hook():
+    """Tests sending and getting back torch nn module like nn.Linear"""
+
+    hook = syft.TorchHook(torch, verbose=True)
```

Comment:
In general we can use the hook from conftest and don't need to rehook for each test. This can be done by passing hook as a parameter to each method. Also it should make the tests run faster

Reply:
Thanks for the feedback. Changes have been made.